Add support for RollForward property

vitek-karas opened the pull request:
The `GenerateRuntimeConfigurationFiles` task now recognizes `RollForward` parameter (and it's passed from the `RollForward` property). It writes its value into the .runtimeconfig.json as the `rollForward` setting.

This implements the SDK side of the roll forward feature as described in the [runtime binding](https://github.com/dotnet/designs/blob/master/accepted/runtime-binding.md#rollforward).

vitek-karas committed:
Add support for RollForward property

The `GenerateRuntimeConfigurationFiles` task now recognizes `RollForward` parameter (and it's passed from the `RollForward` property). It writes its value into the .runtimeconfig.json as the `rollForward` setting.

vitek-karas commented:
@nguerrera Is it custom to validate the allowed values for the new property as part of the build? For example if I would specify `<RollForward>unknown</RollForward>` the produced application would fail to start as `unknown` is not a valid value for the roll forward setting. Do we try to validate this in the SDK itself and fail the build, or do we let this through?

nguerrera commented:
Great question. For all the other enum like properties I can think of, there is validation, but it follows from how they're consumed in the build. For example, OutputType and DebugType. Pass through is not an option there, the build has to bail if it doesn't understand the option.

Here it is used in the build only to write a string, so we could pass through. The benefit of pass through is of course that you could add more options without changing the sdk. However, I don't think that is likely to be a big concern. There's some value in catching the error in the build. I'd say it's better to validate, but I don't have a strong opinion on it. I assume the runtime error should be clear enough, just later than ideal.

@dsplaisted, @peterhuene, thoughts?

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/RuntimeOptions.cs:
```diff
@@ -11,6 +11,9 @@ internal class RuntimeOptions
         public string tfm { get; set; }
 
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public string rollForward { get; set; }
```
Nit: Pre-existing, but I'm not sure why some things are camelCased and other things are PascalCased. See tfm vs Framework(s). It seems that everything will end up camelCased regardless as we're passing some CamelCase thing to JSON.NET? If so, I'd prefer that we used .NET PascalCased convention throughout.

vitek-karas replied:
Good point - didn't notice the magical converter.
I've fixed `tfm` -> `TFM` as well. The camel case conversion seems to convert that to `tfm`.. interesting, but it works.

nguerrera replied:
nit: I would go with Tfm. Framework design guideline for acronyms with more than two letters. E.g. System.Xml, not System.XML.

vitek-karas committed:
PR feedback

vitek-karas committed:
PR feedback

dsplaisted commented:
What should happen if you specify this property for a project that targets .NET Core 2.x?  I think it would only apply at runtime if there is a 3.x or higher runtime installed which would deliver the later host.

Should we generate a warning or error in this case?

vitek-karas committed:
Better naming

vitek-karas commented:
That a good question @dsplaisted  - do we have a precedent for either way?
I would think a warning makes sense in that case.

Technical question - what is the right way to test for 3.0 TFM and above?

dsplaisted commented:
We have precedence for generating errors with invalid combinations of settings.  See the `NETSdkError` task.  The interesting thing here is that there might be advanced scenarios where you want to set this property in a 2.x app where it's not supported.

As for how to do it, the condition would look something like this (to check if the version is less than 3.0):

```
Condition="('$(TargetFrameworkIdentifier)' == '.NETCoreApp') And ('$(_TargetFrameworkVersionWithoutV)' != '') And ('$(_TargetFrameworkVersionWithoutV)' &lt; '3.0')"
```

nguerrera commented:
This might be better in a task too. These conditions in xml have gotten unwieldy IMHO. One thing I noticed recently is that the runtime config generation takes the non-canonical short name for tfm: TargetFramework instead of TargetFrameworkIdentifier and TargetFrameworkVersion. What does the host do with this TFM?

vitek-karas commented:
The host doesn't understand that string in any way. It only uses it to compute probing paths. For example it's possible to specify probing path as `somepath/|arch|/|tfm|` in which case we substitute the `|tfm|` for the value from `.runtimeconfig.json`. There are also cases with global store where the path is computed using the TFM value.

vitek-karas committed:
RollForward validation

Only allow its usage for 3.0 apps.
Validate that the value is one of the supported setting values.

vitek-karas commented:
I decided that it's better to be strict now:
* It's always possible to remove some of the restrictions - going the other way is much harder.
* There is a workaround if necessary using the `.runtimconfig.template` which will bypass any checks (other than the runtime).

I implemented the version check in MSBuild since lot of other features do it that way. But if C# is preferable I can change that (assuming I can figure out how to access the right version numbers).

nguerrera commented on src/Tests/Microsoft.NET.Build.Tests/GivenFrameworkReferences.cs:
```diff
@@ -192,6 +193,78 @@ public void TargetingPackDownloadCanBeDisabled()
                 .HaveStdOutContaining("NETSDK1073");
         }
 
+        [CoreMSBuildOnlyTheory]
```
```suggestion
        [Theory]
```

I'm not seeing why this requires .NET Core msbuild. I think it should pass fine when using .NET Framework msbuild as well. Copied from above?

nguerrera commented on src/Tests/Microsoft.NET.Build.Tests/GivenFrameworkReferences.cs:
```diff
@@ -192,6 +193,78 @@ public void TargetingPackDownloadCanBeDisabled()
+            }
+        }
+
+        [CoreMSBuildOnlyFact]
```
```suggestion
        [Fact]
```

Same here

vitek-karas committed:
PR feedback
